long-ride-alerts/src/main/scala/org/apache/flink/training/exercises/longrides/scala/LongRidesExercise.scala

```diff
@@ -18,56 +18,82 @@
 import org.apache.flink.streaming.api.functions.KeyedProcessFunction
+import org.apache.flink.streaming.api.functions.sink.{PrintSinkFunction, SinkFunction}
+import org.apache.flink.streaming.api.functions.source.SourceFunction
 import org.apache.flink.streaming.api.scala.{StreamExecutionEnvironment, _}
 import org.apache.flink.training.exercises.common.datatypes.TaxiRide
 import org.apache.flink.training.exercises.common.sources.TaxiRideGenerator
-import org.apache.flink.training.exercises.common.utils.ExerciseBase._
-import org.apache.flink.training.exercises.common.utils.{ExerciseBase, MissingSolutionException}
+import org.apache.flink.training.exercises.common.utils.MissingSolutionException
 import org.apache.flink.util.Collector
 
+import java.time.Duration
+
 /**
-  * The "Long Ride Alerts" exercise of the Flink training in the docs.
+  * The "Long Ride Alerts" exercise.
   *
-  * The goal for this exercise is to emit START events for taxi rides that have not been matched
-  * by an END event during the first 2 hours of the ride.
+  * <p>The goal for this exercise is to emit the rideIds for taxi rides with a duration of more than
+  * two hours. You should assume that TaxiRide events can be lost, but there are no duplicates.
   *
+  * <p>You should eventually clear any state you create.
   */
 object LongRidesExercise {
 
-  def main(args: Array[String]) {
+  class LongRidesJob(source: SourceFunction[TaxiRide], sink: SinkFunction[Long]) {
+
+    /**
+      * Creates and executes the ride cleansing pipeline.
+      */
+    @throws[Exception]
+    def execute(): Unit = {
+      val env = StreamExecutionEnvironment.getExecutionEnvironment
+
+      // start the data generator
+      val rides = env.addSource(source)
+
+      // the WatermarkStrategy specifies how to extract timestamps and generate watermarks
+      val watermarkStrategy = WatermarkStrategy
+        .forBoundedOutOfOrderness[TaxiRide](Duration.ofSeconds(60))
+        .withTimestampAssigner(new SerializableTimestampAssigner[TaxiRide] {
+          override def extractTimestamp(ride: TaxiRide, streamRecordTimestamp: Long): Long =
+            ride.getEventTime
+        })
 
-    // set up the execution environment
-    val env = StreamExecutionEnvironment.getExecutionEnvironment
-    env.setParallelism(ExerciseBase.parallelism)
+      // create the pipeline
+      rides
+        .assignTimestampsAndWatermarks(watermarkStrategy)
+        .keyBy(_.rideId)
+        .process(new AlertFunction())
+        .addSink(sink)
 
-    val rides = env.addSource(rideSourceOrTest(new TaxiRideGenerator()))
+      // execute the pipeline
+      env.execute("Long Taxi Rides")
+    }
 
-    val longRides = rides
-      .keyBy(_.rideId)
-      .process(new ImplementMeFunction())
+  }
 
-    printOrTest(longRides)
+  @throws[Exception]
+  def main(args: Array[String]) {
+    val job = new LongRidesJob(new TaxiRideGenerator, new PrintSinkFunction)
 
-    env.execute("Long Taxi Rides")
+    job.execute
```

Comment:
IntelliJ suggests to add parameters if it has side effects (which it does)
```suggestion
    job.execute()
```